### ISSUE TITLE
Feature Request: Allow await in top-level Script code

### ISSUE THREAD

**Issue by SLaks** [dotnet/roslyn#1382]
Can you change the `Script` compiler to put top-level code in a `public async Task Execute() { ... }` to allow `await`s?

This would be very useful when using Roslyn to write scripts to automate async things; users should be able to write both `using` statements (so I can't wrap in my own method) and top-level code.

**Comment by filipw** [comment]
+1

**Comment by khellang** [comment]
:+1:

**Comment by ManishJayaswal** [comment]
C# has implemented everything. We still need to add support for extension methods and await in VB

**Comment by patriksvensson** [comment]
Extension method support in top-level code is something that I would really want :+1:

**Comment by khellang** [comment]
> Extension method support in top-level code is something that I would really want

What does this mean? You want to be able to extend the script wrapper class so you can do `this.ExtensionMethod()`?

**Comment by patriksvensson** [comment]
Yes. Either by `this` and also treating the extension method for the script
host type as a method on the script host.

I.e.

```csharp
// Like this:
this.ExtensionMethod();

// Or this:
ExtensionMethod();
```

**Comment by SLaks** [comment]
You could do that using `using static`.

**Comment by khellang** [comment]
@patriksvensson You could also just have a property on the host object which you can hang extension methods off?

**Comment by patriksvensson** [comment]
Using static loses the context of the script host (unless it would be a
singleton).

Using a property on the script host would of course work, but that's not
what this was about (unless I misunderstood the issue; in that case I
apologize) :-)

**Comment by SLaks** [comment]
True.

As long as scripts don't call other scripts, you could pass the context as a `[ThreadStatic]` field.

**Comment by patriksvensson** [comment]
@SLaks Yes, you're right, that would be one approach. Would be nice if this was supported by Roslyn though. Would be less "hacky" :smile:

**Comment by tmat** [comment]
@patriksvensson As @khellang said, you can define a public method ExtensionMethod() on the host class and it is accessible from the script like so:

```C#
ExtensionMethod()
```

**Comment by patriksvensson** [comment]
@tmat Yes, I could do that. The scenario this would be useful for is DSLs. I run a small build automation system called [Cake](http://cakebuild.net) and we allow extension methods to our script host to be called directly in a script.

So if someone creates an extension method that looks like this:

```csharp
public static void NuGetRestore(this ICakeContext context, FilePath solutionPath)
{
}
```

It can be used in a script like this:

```csharp
Task("DoMagic") // Part of the Cake RunTime
  .Does(() =>       // Also part of the Cake script host
{
    NuGetRestore("./src/MyProject.sln"); // Method specified by third party addin.
});
```

We do this today by generating a proxy method in the script that gets executed, but it would involve much less "magic" if extensionmethods to the script host class could be invoked just like a method on the script host class.

I understand if this is not something you think is necessary in the scripting APIs, but I think it would be something that other projects could benefit from as well. Anyway, this is the reason I want this.

**Comment by tmat** [comment]
@patriksvensson I see. I assume you don't like the script to use an indirection, right? Something like 

```C#
Host.NuGetRestore("./src/MyProject.sln"); 
```

**Comment by khellang** [comment]
I'd stay far away from that API anyway. How do you even provide IntelliSense for that? Wouldn't it be much better to do something like

```
Task("DoMagic") // Part of the Cake RunTime
  .Does(context =>       // Also part of the Cake script host
{
    context.NuGetRestore("./src/MyProject.sln"); // Method specified by third party addin.
});
```

At least you have a discoverable API :smile: This is just general API design though.

**Comment by tmat** [comment]
@khellang You can still get completion (Alt+ArrowRight). I think it's just matter of design preference.

**Comment by khellang** [comment]
@tmat The same as i VS? If I have the caret in an instance method and pull out IntelliSense, it usually gives some horrible results...

**Comment by tmat** [comment]
@khellang Extension methods are not available without "this." prefix. It should give you all the identifiers and keywords that are possible at the current location.

**Comment by khellang** [comment]
Exactly. That's why I'd find it weird if top-level extensions were supported (without `this`). Even with `this` it'd feel weird since you're not in a "`this`" context.

**Comment by tmat** [comment]
@khellang and that's exaclty why they are currently not :)

**Comment by patriksvensson** [comment]
@khellang The context parameter is already supported. The variant I was talking about is syntactic sugar (which 99% of our users seem to prefer).

> I'd stay far away from that API anyway.

You're free to stay far away from that API. Won't force you, promise :wink:
